cg/cg_finch.jl

```diff
@@ -25,60 +25,53 @@ function ssymv_finch(y, A, x)
     y
 end
 
+# Using algorithm from https://github.com/JuliaLinearAlgebra/IterativeSolvers.jl/blob/e5d10e22552916e7dbd73d0a9ece02d873dee375/src/cg.jl
 function cg_finch_kernel(x, A, b, l)
     (n, m) = size(A)
     @assert n == m
 
-    r = Fiber!(Dense(Element(0.0), n))
-    _r = Fiber!(Dense(Element(0.0), n))
-    p = Fiber!(Dense(Element(0.0), n))
-    _p = Fiber!(Dense(Element(0.0), n))
-    _x = Fiber!(Dense(Element(0.0), n))
-    Ap = Fiber!(Dense(Element(0.0), n))
+    _x = Tensor(Dense(Element(0.0)), zeros(n))
+    u = Tensor(Dense(Element(0.0)), zeros(n))
+    _u = Tensor(Dense(Element(0.0)), zeros(n))
+    r = Tensor(Dense(Element(0.0)), zeros(n))
+    _r = Tensor(Dense(Element(0.0)), zeros(n))
+    c = Tensor(Dense(Element(0.0)), zeros(n))
 
-    # r_0 = b - Ax_0
-    ssymv_finch(_r, A, x)
-    @finch for i = _; r[i] = b[i] - _r[i] end
-    # p_0 = r_0
-    @finch for i = _; p[i] = r[i] end
-    # r2 = r_0^Tr_0
-    r2 = Scalar(0.0)
-    @finch (for i = _; r2[] += r[i]^2 end)
+    ssymv_finch(c, A, x)
+    @finch for i = _; r[i] = b[i] - c[i] end
+    residual_sq = Scalar(0.0)
+    @finch (for i = _; residual_sq[] += r[i]^2 end)
+    residual = Scalar((residual_sq[])^(1/2))
 
+    prev_residual = Scalar(1.0)
     for k = 1:l
-        pTAp = Scalar(0.0)
+        β = Scalar(residual[]^2 / prev_residual[]^2)
+        @finch (for i = _; _u[i] = r[i] + β[] * u[i] end)
```

Comment:
Could you try updating some of these in-place? If it doesn't work that's okay but I think it's worth a shot, especially if u is dense

Reply:
I don't think this suggestion is well-founded after all. Finch can't handle that kind of update.

Reply:
Okay got it—that's what I thought. I was confused how this would work, but I was trying it out anyways (and yes Finch was not able to handle the in-place updates)